enhancements/20240523-support-rwx-volume-to-downstream-cluster.md

```diff
@@ -0,0 +1,110 @@
+
+Reference: https://www.mongodb.com/docs/kubernetes-operator/v1.25/tutorial/configure-file-store/
+
+### User Experience In Detail
+
+With this feature, the user could create volumes with two options in the downstream cluster. The first is the RWO volume. The second is the RWX volume. Then user could attach use the RWX volume as previous RWO volume. The Harvester CSI driver would handle the internal work.
+The UX did not change too much. The only difference is the user needs to select the RWX volume.
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A quick overview of the design could be checked below:
+
+![Downstream RWX Support](20240523-support-rwx-volume-to-downstream-cluster/general-rwx.png)
+
+### Solution 1.
+
+Like the RWO solution, we would like to leverage the libvirt capability to hotplug the filesystem and then use the workload pod as a host mountpoint, but currently, it is not supported. The hotplug filesystem on the kubevirt side is referred to as virtiofs. However, virtiofs does not support live migration, so we could not use this solution.
+
+Related feature request to Kubevirt: https://github.com/kubevirt/kubevirt/issues/11699
+
+### Solution 2.
+
+The second solution is like the Longhorn CSI behavior with RWX volume. The Longhorn share manager will provide the NFS endpoint through ClusterIP. The thing we need to take care of is network space. The downstream cluster uses a different network space from the Harvester/Longhorn cluster. We need to ensure the downstream cluster can interact with the NFS endpoint with the same network space. Like below:
+
+![Network Space](20240523-support-rwx-volume-to-downstream-cluster/rwx-network.png)
+
+The detailed flowchart is like below:
+
+- RWX volume Create:
+
+![RWX volume Create](20240523-support-rwx-volume-to-downstream-cluster/create-rwx-volume.png)
+
+- RWX volume Attachment:
+
+![RWX volume Attachment](20240523-support-rwx-volume-to-downstream-cluster/attach-rwx-volume.png)
+
+### Each component detailed design
+
+#### 1. Longhorn
+Longhorn share-manager can provide the NFS endpoint through the storage network.
+
+Related issue: https://github.com/longhorn/longhorn/issues/8184
+
+#### 2. Harvester CSI Driver
+Harvester CSI Driver need to add capability to create the RWX volume through Longhorn client. Also, we should add the corresponding handles on NodeStage/NodePublish stage. The NodeStage will mount the NFS endpoint to the host as global path. Then, the NodePublish will bind mount the global path for the workload.
+
+The whole flow might be like that when get the request to create RWX volume, Harvester CSI Driver need to create corresponding RWX volume on Longhorn. Then, the Harvester CSI driver will create a global path on NodeStage. Last, with NodePublish, the Harvester CSI driver will bind mount the global path for the workload.
```

Comment:
> Harvester CSI Driver need to create corresponding RWX volume on Longhorn

Is this limited to LH only? What will it take to make this more generic? Theoretically, the Harvester CSI driver can create RWX volumes of any storage class, and it's up to the Harvester-side CSI drivers to do the right thing, right?